### PR TITLE
Always trigger guard condition waitset

### PR DESCRIPTION
**Pull request by alsora**
Ttrigger guard condition waitset regardless of whether a trigger callback is present
This should address https://github.com/ros2/rclcpp/issues/1917

Also minor cleanup at the class (remove unneded lock and better use of if-else)

Signed-off-by: Alberto Soragna <alberto.soragna@gmail.com>

### CODE REVIEW

File: rclcpp/src/rclcpp/guard_condition.cpp [hunk 1]
```diff
@@ -96,8 +99,6 @@ GuardCondition::exchange_in_use_by_wait_set_state(bool in_use_state)
 void
 GuardCondition::add_to_wait_set(rcl_wait_set_t * wait_set)
 {
-  std::lock_guard<std::recursive_mutex> lock(reentrant_mutex_);
-
```
Comment (fujitatomoya) [inline rclcpp/src/rclcpp/guard_condition.cpp:100]: How about the following MT case?

1. Thread-A comes with `wait_set-X`.
2. Context switch to thread-B right after thread-A stores `wait_set-X` to `wait_set_`.
3. Thread-B also comes with `wait_set-X`.
4. Thread-B and Thread-A might call `rcl_wait_set_add_guard_condition` at the same time.

as far as i see the doc of `rcl_wait_set_add_subscription`, these functions are not thread-safe?

Comment (alsora) [reply rclcpp/src/rclcpp/guard_condition.cpp:100]: I'm not sure.
The other waitables (e.g. actions, intra-process sub) do not use a mutex there.
Also the executor does not seem to use a mutex when adding entities (subscriptions, services, etc) to waitsets.

Comment (wjwwood) [reply rclcpp/src/rclcpp/guard_condition.cpp:100]: It's undefined behavior to add an item to two different wait sets, and it's also undefined behavior to try to wait on a single wait set from two different threads.

There is a thread-safe `rclcpp::WaitSet` that protects this case from happening.

Comment (alsora) [reply rclcpp/src/rclcpp/guard_condition.cpp:100]: It looks to me that the `add_to_wait_set`  function is already preventing to add an item to two different wait-sets.

This is the current code.
The atomic `in_use_by_wait_set_` should be enough to avoid the undefined behavior, without need for a mutex.

```
void
GuardCondition::add_to_wait_set(rcl_wait_set_t * wait_set)
{
  if (exchange_in_use_by_wait_set_state(true)) {
    if (wait_set != wait_set_) {
      throw std::runtime_error("guard condition has already been added to a wait set.");
    }
  } else {
    wait_set_ = wait_set;
  }

  rcl_ret_t ret = rcl_wait_set_add_guard_condition(wait_set, &this->rcl_guard_condition_, NULL);
  if (RCL_RET_OK != ret) {
    rclcpp::exceptions::throw_from_rcl_error(
      ret, "failed to add guard condition to wait set");
  }
}
```

Comment (alsora) [reply rclcpp/src/rclcpp/guard_condition.cpp:100]: It looks to me that the `add_to_wait_set`  function is already preventing to add an item to two different wait-sets.

This is the current code.
The atomic `in_use_by_wait_set_` should be enough to avoid the undefined behavior, without need for a mutex.

```
void
GuardCondition::add_to_wait_set(rcl_wait_set_t * wait_set)
{
  if (exchange_in_use_by_wait_set_state(true)) {
    if (wait_set != wait_set_) {
      throw std::runtime_error("guard condition has already been added to a wait set.");
    }
  } else {
    wait_set_ = wait_set;
  }

  rcl_ret_t ret = rcl_wait_set_add_guard_condition(wait_set, &this->rcl_guard_condition_, NULL);
  if (RCL_RET_OK != ret) {
    rclcpp::exceptions::throw_from_rcl_error(
      ret, "failed to add guard condition to wait set");
  }
}
```

**Comment by alsora** [comment]
I restored the mutex to prevent race condition that could cause problems.

File: rclcpp/src/rclcpp/guard_condition.cpp [hunk 2]
```diff
@@ -97,7 +100,7 @@ void
 GuardCondition::add_to_wait_set(rcl_wait_set_t * wait_set)
 {
   std::lock_guard<std::recursive_mutex> lock(reentrant_mutex_);
-
+
```
Comment (fujitatomoya) [inline rclcpp/src/rclcpp/guard_condition.cpp:103]: nit: unnecessary difference.

**Review by fujitatomoya — APPROVED** [review]
lgtm with minor nit

**Review by wjwwood — COMMENTED** [review]
Looks reasonable.

It seems like a test to avoid this regressing in the future would be good to have.

Also, if it's not already clear from the docstring what should happen in this corner case, it should be clarified there.

**Comment by alsora** [comment]
Added unit-tests.

Running full CI

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/16719/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/11299/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/17113/)

**Comment by wjwwood** [comment]
This doc block should mention how it interacts with the "on trigger callback":

https://github.com/ros2/rclcpp/blob/02802bcc385c3d6d814add93618f3388d44adec7/rclcpp/include/rclcpp/guard_condition.hpp#L75-L85

And, I guess we missed it, but we should have some documentation on the "on trigger callback" method as well (it currently has none):

https://github.com/ros2/rclcpp/blob/02802bcc385c3d6d814add93618f3388d44adec7/rclcpp/include/rclcpp/guard_condition.hpp#L110-L112

**Comment by alsora** [comment]
@wjwwood @fujitatomoya anything else to do here?

**Review by wjwwood — APPROVED** [review]
lgtm

Sorry for not re-reviewing it. I don't get notifications for pushed commits, so a comment is helpful to bump a re-review.
